src/rpc/methods/eth/filter/mod.rs

```diff
@@ -0,0 +1,263 @@
+            max_height = max;
+        }
+
+        for ea in &self.address {
+            let addr = ea
+                .to_filecoin_address()
+                .map_err(|e| format!("invalid address {}", e))?;
+            addresses.push(addr);
+        }
+
+        let keys = parse_eth_topics(&self.topics)?;
+
+        Ok(ParsedFilter {
+            min_height,
+            max_height,
+            tipset_cid,
+            addresses,
+            keys: keys_to_keys_with_codec(keys),
+        })
+    }
+}
+
+fn parse_block_range(
+    heaviest: ChainEpoch,
+    from_block: Option<&str>,
+    to_block: Option<&str>,
+    max_range: ChainEpoch,
+) -> Result<(ChainEpoch, ChainEpoch), String> {
+    let min_height = match from_block {
+        None | Some("latest") | Some("") => heaviest,
+        Some("earliest") => 0,
+        Some(block) => {
+            if !block.starts_with("0x") {
+                return Err("FromBlock is not a hex".to_string());
+            }
+            hex_str_to_epoch(block).map_err(|_| "invalid epoch".to_string())?
+        }
+    };
+
+    let max_height = match to_block {
+        None | Some("latest") | Some("") => -1,
+        Some("earliest") => 0,
+        Some(block) => {
+            if !block.starts_with("0x") {
+                return Err("ToBlock is not a hex".to_string());
+            }
+            hex_str_to_epoch(block).map_err(|_| "invalid epoch".to_string())?
+        }
+    };
+
+    if min_height == -1 && max_height > 0 {
+        if max_height - heaviest > max_range {
+            return Err(format!(
+                "invalid epoch range: to block is too far in the future (maximum: {})",
+                max_range
+            ));
+        }
+    } else if min_height >= 0 && max_height == -1 {
+        if heaviest - min_height > max_range {
+            return Err(format!(
+                "invalid epoch range: from block is too far in the past (maximum: {})",
+                max_range
+            ));
+        }
+    } else if min_height >= 0 && max_height >= 0 {
+        if min_height > max_height {
+            return Err(format!(
+                "invalid epoch range: to block ({}) must be after from block ({})",
+                max_height, min_height
+            ));
+        } else if max_height - min_height > max_range {
+            return Err(format!(
+                "invalid epoch range: range between to and from blocks is too large (maximum: {})",
+                max_range
+            ));
+        }
+    }
+
+    Ok((min_height, max_height))
+}
+
+fn hex_str_to_epoch(hex_str: &str) -> Result<ChainEpoch, String> {
+    let hex_substring = hex_str
+        .get(2..)
+        .ok_or_else(|| "invalid hex string: unable to parse epoch".to_string())?;
+    i64::from_str_radix(hex_substring, 16).map_err(|e| e.to_string())
+}
+
+fn parse_eth_topics(topics: &EthTopicSpec) -> Result<HashMap<String, Vec<Vec<u8>>>, String> {
+    let mut keys: HashMap<String, Vec<Vec<u8>>> = HashMap::new();
```

Comment:
Do we know in advance how many keys will be there?

Reply:
Each eth log entry can contain up to 4 topics, corrected